src/main/scala/chisel3/util/MixedVec.scala

```diff
@@ -0,0 +1,112 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3._
+import chisel3.core.{Data, requireIsChiselType, requireIsHardware}
+import chisel3.internal.naming.chiselName
+
+import scala.collection.immutable.ListMap
+
+object MixedVecWireInit {
+  /**
+    * Construct a new wire with the given bound values.
+    * This is analogous to [[chisel3.core.VecInit]].
+    * @param vals Values to create a MixedVec with and assign
+    * @return MixedVec with given values assigned
+    */
+  def apply[T <: Data](vals: Seq[T]): MixedVec[T] = {
+    // Create a wire of this type.
+    val hetVecWire = Wire(MixedVec(vals.map(_.cloneTypeFull)))
+    // Assign the given vals to this new wire.
+    for ((a, b) <- hetVecWire.zip(vals)) {
+      a := b
+    }
+    hetVecWire
+  }
+}
+
+object MixedVec {
+  /**
+    * Create a MixedVec from the type of the given Vec.
+    * For example, given a Vec(2, UInt(8.W)), this creates MixedVec(Seq.fill(2){UInt(8.W)}).
+    * @param vec Vec to use as template
+    * @return MixedVec analogous to the given Vec.
+    */
+  def apply[T <: Data](vec: Vec[T]): MixedVec[T] = {
+    MixedVec(Seq.fill(vec.length)(vec.sample_element))
+  }
+}
+
+/**
+  * A hardware array of elements that can hold values of different types/widths,
+  * unlike Vec which can only hold elements of the same type/width.
+  *
+  * @param eltsIn Element types. Must be Chisel types.
+  *
+  * @example {{{
+  * val v = Wire(HeterogeneousVec(Seq(UInt(8.W), UInt(16.W), UInt(32.W))))
+  * v(0) := 100.U(8.W)
+  * v(1) := 10000.U(16.W)
+  * v(2) := 101.U(32.W)
+  * }}}
+  */
+@chiselName
+final case class MixedVec[T <: Data](private val eltsIn: Seq[T]) extends Record with collection.IndexedSeq[T] {
+  // We want to create MixedVec only with Chisel types.
+  if (compileOptions.declaredTypeMustBeUnbound) {
+    eltsIn.foreach(e => requireIsChiselType(e))
+  }
+
+  // Clone the inputs so that we have our own references.
+  private val elts: IndexedSeq[T] = eltsIn.map(_.cloneTypeFull).toIndexedSeq
+
+  /**
+    * Statically (elaboration-time) retrieve the element at the given index.
+    * @param index Index with which to retrieve.
+    * @return Retrieved index.
+    */
+  def apply(index: Int): T = elts(index)
+
+  /**
+    * Dynamically (via a mux) retrieve the element at the given index.
+    *
+    * @param index Index with which to retrieve.
+    * @return Retrieved index.
+    */
+  def apply(index: UInt): T = {
+    requireIsHardware(index, "index must be hardware")
+
+    if (length < 1) {
+      throw new IndexOutOfBoundsException("Collection is empty")
+    }
+
+    MuxLookup(index, elts.head, elts.zipWithIndex.map { case (el, el_index) => el_index.U -> el })
```

Comment:
This actually won't work, you can't create a mux with heterogeneous types. This is a big problem but it's not really clear how you can possibly handle dynamically indexing heterogeneous types...

Reply:
Hmm, would it be possible for you to elaborate on that? I wrote a test for how I would use it from a user's perspective (MixedVecDynamicIndexTester) and it seems to behave/work as expected.

Reply:
The issue with the test is that it only checks dynamic indexing when all of the elements are `UInts`, if you try dynamically indexing a `MixedVec` composed of a `UInt` and some other `Data` like `SInt` or `Bundle`, you will get an elaboration time exception

Reply:
Thanks for the catch Jack. I thought about it some more, and came up with a solution that seems to work for heterogeneous elements as long as the user is careful about it, and added a unit test to demonstrate/test it.

Since Chisel's mux already creates a result of the maximum width (e.g. muxing `UInt(3.W)` and `UInt(8.W)` -> `UInt(8.W)`), so we just return a UInt of the maximum width and let the designer handle packing/unpacking/etc. I think this interpretation makes the most sense in light of dynamic indexing as a hardware construction operation.